### ISSUE TITLE
[TPU] jax 0.10.2 -> 0.11.1: +57% decode step-time regression on v7x from SparseCore gather offload at batch=1; recovered by --xla_tpu_offload_gather_to_sparsecore=false

### ISSUE THREAD

**Issue by zorrofox** [jax-ml/jax#40553]
Upgrading jax[tpu] 0.10.2 -> 0.11.1 regresses single-stream decode step time for a 78-layer MoE model (GLM-5.2: MLA + sparse attention, 256 experts top-k 8, tp16 across 2 hosts on TPU v7x, 135k-context serving config, sglang-jax) from **13.7 ms/token to 21.4 ms/token** (+57%, fresh-boot probe; a warm/reloaded server measures 17.6 ms, see note on state dependence below). Same-node same-hour A/B with isolated compilation caches; 0.10.2 reproduces its 5-day-old baseline within 0.3%, so environment and hardware are controlled.

**Root cause (confirmed by trace + flag ablation): 0.11.1's libtpu offloads gathers in the bs=1 decode graph to SparseCore, and for these shapes the offload is a large net loss.** The 0.11.1 trace of the same workload shows `gather_offload` (~5.3 ms/step) plus `branch_1_fun` (~1.7 ms/step) that do not exist in the 0.10.2 trace. Setting `LIBTPU_INIT_ARGS="--xla_tpu_offload_gather_to_sparsecore=false --xla_tpu_offload_all_supported_gathers_to_sparsecore=false"` makes both ops disappear from the trace and recovers decode step time completely. Measurement matrix (all on 0.11.1 unless noted, TPOT via decode-only differential timing (t(max_new=257)-t(max_new=1))/256, temperature 0, ignore_eos, two rounds each, spread < 0.1%):

| code | jax | LIBTPU_INIT_ARGS | ms/token |
|---|---|---|---|
| commit A | 0.10.2 | none | 13.67 / 13.76 |
| commit B (later) | 0.11.1 | none | 21.46 |
| commit B | 0.11.1 | 2 flags above | **13.66 / 13.66** |
| commit C (=B plus unrelated changes) | 0.11.1 | 2 flags above | 14.60 / 14.60 |
| commit C | 0.11.1 | 8 flags (all SC offload paths off) | 14.54 / 14.53 |

Commit B on 0.11.1 with the two flags exactly matches the 0.10.2 baseline (13.66 vs 13.67), i.e. the flags recover 100% of the version regression; the commit C rows show the 2-flag set is equivalent to disabling every SparseCore offload path (delta 0.06 ms, within noise), so the gather-offload pass alone is the regression.

**The offload is genuinely profitable for large gathers on the same setup, so a global disable is not a viable workaround.** 110k-token chunked prefill (chunk 8192): commit C prefills in 24.3 s with default offload but 29.3 s (+21%) with the two flags; commit B similarly goes 30.0 s -> 36.2 s. Notably commit B's prefill on 0.11.1 with default offload (30.0 s) matches its 0.10.2 baseline (30.2 s) while its decode regresses +57% — consistent with the "auto" heuristic (public IR dumps show `xla_tpu_offload_gather_to_sparsecore` defaults to `auto`) making the right call for large prefill gathers and the wrong call for tiny bs=1 decode gathers. We verified a scoped workaround: the flag is honored per-executable via `jax.jit(f).lower(...).compile(compiler_options={"xla_tpu_offload_gather_to_sparsecore": "false"})` — the offload forms disappear from that executable's optimized HLO only — so a serving framework can disable it for the decode executable while keeping it for prefill.

Three secondary observations that may help:

1. **Step time on 0.11.1 (without the flags) depends on server history** — ~21.4 ms/token probed from a fresh boot vs ~17.6 ms/token probed after a sequence of heavy 110k-token prefills, both individually stable to +/-0.1 ms. On 0.10.2 both probe sequences read the same 13.7-13.8 ms. Whatever decides the offload's cost appears to be stateful.

2. **Stability**: on 0.11.1 the first server launch crashed with SIGABRT (exit 134, "Fatal Python error: Aborted") during early compilation in 4 independent bring-ups across two node pairs, succeeding on relaunch; never observed on 0.10.2.

3. **Prefill also regresses on the same upgrade, independently of this offload issue**: with default flags, 110k-token TTFT goes 22.1 s -> 24.3 s (+10%) on our optimized kernel path, while an older code revision without those kernel optimizations is version-neutral (30.2 s -> 30.0 s) — i.e. 0.11.1 specifically eats part of the gain of custom prefill kernels. We are still isolating this to a specific op and will file it separately once we have a trace-level attribution; this issue is scoped to the decode gather-offload regression, which is fully characterized above.

Questions: (a) what does the `auto` heuristic evaluate on this TPU generation, and what changed between the libtpu builds bundled with 0.10.2 and 0.11.1 that makes it now offload these decode-graph gathers (it looks like previously-rejected shapes, e.g. index counts not aligned to SparseCore lane width, are now accepted)? (b) is there a supported per-op annotation or a size/duplication threshold to keep the offload for large gathers but reject small ones, rather than the all-or-nothing flag? (c) should the cost of the offload (cross-core sync + the sort inserted before the gather) be modeled so `auto` rejects cases where it cannot pay off, like bs=1 decode? We are happy to run candidate flags or patched libtpu builds on v7x and can provide full xplane traces for both versions on request.

**Comment by zorrofox** [comment]
Follow-up with op-level attribution and a scoped-workaround verification.

**The offloaded gather** (from the 135k-config decode trace) is a sparse-KV collection: `gather_offload(operand bf16[2332,128,128] paged-KV pool ~76 MB, indices s32[2048]) -> bf16[1056,128,128]`, async on `execution_thread="sparsecore"`, `offload_duration ~0.7 ms` but ~5 ms/step spent in the async-done wait. The producer is a scatter (KV write-back) and the consumer is our attention kernel.

**Why only long-context configs are hit**: the same model at a 49k context config (same 2048-index gather, KV pool ~13 MB) is completely unaffected — TPOT identical with and without the flags at every concurrency level. So the offload decision appears to key on the operand not fitting in VMEM, and the regression scope is "sparse/paged-KV decode with an HBM-resident table", i.e. long-context serving.

**Per-executable scoping works end-to-end**: compiling only the decode-shaped executable with `compiler_options={"xla_tpu_offload_gather_to_sparsecore": "false", "xla_tpu_offload_all_supported_gathers_to_sparsecore": "false"}` (via lower().compile in our serving framework) recovers decode fully (21.4 -> 14.5 ms/token) while prefill keeps the offload benefit (110k TTFT stays 24.4 s vs 29.3-29.9 s with the process-global flags). This is a usable workaround for serving frameworks; still, a cost model (or size/duplication heuristic) that rejects the offload when the per-step win cannot cover the cross-core sync + inserted sort would fix this for everyone at batch=1.

Also for the record: the first-launch SIGABRT mentioned above has now reproduced 4 times across two node sets, always succeeding on relaunch.

**Comment by luispreciado-byte** [comment]
Thank you for an extraordinarily well-characterized report

### Clarification on Gather Offload Heuristic, Workarounds, and Next Steps.

### Why only 135k context regresses while 49k context is unaffected

When a table is small enough to stay resident on-chip, offloading to SparseCore is skipped to favor low-latency native TensorCore execution.

### Clarifications on Your Questions

#### (a) What does `auto` evaluate to on TPU v7x, and what changed in 0.11.1?
* **The `auto` heuristic:** At the flag level (`xla_tpu_offload_gather_to_sparsecore=auto`), `auto` functions as an architectural enablement switch rather than an instruction-level cost model. For TPU v7x, `auto` enables offloading capability for qualifying gather operations.

### What changed between the libtpu builds bundled with 0.10.2 and 0.11.1 that makes it now offload these decode-graph gathers?

In 0.11.1 the XLA compiler.introduced support for offload gathers with arbitrary/non-divisible index counts. While the reported shape is s32[2048] which passes the compiler’s divisibility constraint. The changes affect the lowering stages which also affect runtime performance.

#### (b) Supported per-op annotations, thresholds, and recommended workarounds

* **Per-executable compilation options (Recommended):** Your verified workaround using per-executable compiler options:

```python
      compiler_options={
          "xla_tpu_offload_gather_to_sparsecore": "false",
          "xla_tpu_offload_all_supported_gathers_to_sparsecore": "false",
      }
```

passed via``` jax.jit(f).lower(...).compile(compiler_options=...)``` is officially supported and a viable workaround.

#### (b) Supported per-op annotations or size thresholds to reject small gathers

Per-Op Annotation via compute_on:

Any gather marked with dense compute is explicitly rejected from SparseCore offload. In JAX, you can annotate specific operations using:

```
from jax.experimental import compute_on
    
    pinned_gather = compute_on(
        gather_fn,
        compute_type="device",
        out_memory_spaces=jax.memory.Space.Device,
    )
```

#### (c) Modeling the offload cost (cross-core sync + inserted sort)

You are completely right. Currently, the XLA compiler lacks a robust execution cost model to determine if a gather to offload is profitable:

We have tracked this issue internally and are actively working on improving the cost model.

### Request for Traces

Could you please attach the Xplane traces for both the 0.10.2 baseline and the 0.11.1 run? Having the exact timeline traces showing will be immensely helpful in confirming the root cause.

**SIGABRT during compilation:** Exit code 134 indicates an abort/assertion failure in the compiler backend. If you have the terminal output or traceback showing the crash stack, please share it so we can investigate the root cause.

**Comment by zorrofox** [comment]
Here are the requested Xplane traces (hosted as release assets, links below), both captured with the same methodology: 20 decode steps at batch=1 on the 135k-context config (1024-token prompt, then profiled mid-generation via the serving framework's profiler), same model, same node pool. To fit the attachment limit each trace is trimmed to the `/device:TPU:0` (+ SparseCore 0) planes; the full 16-chip captures (~80 MB gzipped each) are available if useful.

- `jax0102-baseline-decode`: jax/libtpu 0.10.2, decode ~13.7 ms/token, no `gather_offload` ops.
- `jax0111-regression-decode`: jax/libtpu 0.11.1 (same code revision family, no workaround flags), decode ~21 ms/token; each step carries the `gather_offload_async_start/update/done` chain on `execution_thread="sparsecore"` with `offload_duration ~0.7 ms` and ~5 ms/step spent in the async-done wait, plus the `branch_1_fun` sort branch.

On the SIGABRT: it has not reproduced in the last several bring-ups (it appears tied to fully-cold parallel compilation; recent restarts hit the persistent compile cache). We will capture and attach the crash stack the next time it fires on a cold cache.

Also confirming your `compute_on` suggestion is noted — for now we shipped the per-executable `compiler_options` route in our serving framework (sgl-project/sglang-jax#1631) since it needs no model-code annotations; per-op pinning is a good follow-up once the cost model lands.

Downloads:
- [jax0102-baseline-decode-tpu0only.xplane.pb.gz](https://github.com/zorrofox/sglang-jax/releases/download/traces-jax40553/jax0102-baseline-decode-tpu0only.xplane.pb.gz) (13.1 MB)
- [jax0111-regression-decode-tpu0only.xplane.pb.gz](https://github.com/zorrofox/sglang-jax/releases/download/traces-jax40553/jax0111-regression-decode-tpu0only.xplane.pb.gz) (13.4 MB)
